src/jit/earlyprop.cpp

```diff
@@ -296,6 +296,33 @@ bool Compiler::optEarlyPropRewriteTree(GenTreePtr tree)
                 // array length argument, but the type of GT_ARR_LENGTH is always INT32.
                 return false;
             }
+
+            // When replacing GT_ARR_LENGTH nodes with constants we can end up with GT_ARR_BOUNDS_CHECK
+            // nodes that have constant operands and thus can be trivially proved to be useless. It's
+            // better to remove these range checks here, otherwise they'll pass through assertion prop
+            // (creating useless (c1 < c2)-like assertions) and reach RangeCheck where they are finally
+            // removed. Common patterns like new[] { 1, 2, 3 } benefit from this.
+
+            if ((tree->gtNext != nullptr) && tree->gtNext->OperIs(GT_ARR_BOUNDS_CHECK))
+            {
+                GenTreeBoundsChk* check = tree->gtNext->AsBoundsChk();
```

Comment:
Worth an assert that `check->gtArrLen == tree` ?

Reply:
Oops, definitely. Or even a proper `if` check, it's not worth the risk.